File: 09-logicno-programiranje/stevila.ml
(* Naravna števila z golimi rokami. *)

type nat =
  | Zero
  | Succ of nat

(* Primer *)
let three = Succ (Succ (Succ Zero))

(* Seštevanje *)
let rec plus = function
  | (x, Zero) -> x
  | (x, Succ y) -> Succ (plus (x, y))

(* Primer *)
let five = plus (three, Succ (Succ Zero))

File: 09-logicno-programiranje/stevila.pl
nat(z).
nat(s(X)) :- nat(X).

plus(X, z, X).
plus(X, s(Y), s(Z)) :- plus(X, Y, Z).

times(_, z, z).
times(X, s(Y), Z) :-
    times(X, Y, W),
    plus(X, W, Z).
